application/comit_node/src/comit_client/bam.rs

```diff
@@ -150,69 +151,141 @@ fn to_json_value(
-        debug!("Trying to get client for {}", comit_node_socket_addr);
-        let existing_client = self
-            .clients
-            .read()
-            .unwrap()
-            .get(&comit_node_socket_addr)
-            .cloned();
-
-        match existing_client {
-            None => {
+        socket_addr: SocketAddr,
+    ) -> Box<dyn Future<Item = Result<Arc<BamClient>, ClientFactoryError>, Error = ()> + Send> {
+        debug!("Trying to get client for {}", socket_addr);
+        let mut clients = self.clients.write().unwrap();
+
+        let existing_connection = clients
+            .entry(socket_addr)
+            .or_insert_with(|| Arc::new(Mutex::new(ConnectionState::Disconnected)));
+
+        let mut connection_state = (*existing_connection).lock().unwrap();
+
+        match *connection_state {
+            ConnectionState::Disconnected => {
                 info!(
                     "No existing connection to {}. Trying to connect.",
-                    comit_node_socket_addr
+                    socket_addr
                 );
-                let socket = TcpStream::connect(&comit_node_socket_addr).wait()?;
-                info!("Connection to {} established", comit_node_socket_addr);
-                let codec = json::JsonFrameCodec::default();
-
-                let response_source = Arc::new(Mutex::new(json::JsonResponseSource::default()));
-                let incoming_frames = json::JsonFrameHandler::create(
-                    Config::<json::Request, json::Response>::default(),
-                    Arc::clone(&response_source),
-                );
-                let (client, outgoing_frames) = bam::client::Client::create(response_source);
-                let connection = connection::new(codec, socket, incoming_frames, outgoing_frames);
-                let socket_addr = comit_node_socket_addr;
-                tokio::spawn(connection.map_err(move |e| {
-                    error!(
-                        "Connection to {:?} prematurely closed: {:?}",
-                        socket_addr, e
-                    )
-                }));
-                let client = Arc::new(BamClient::new(comit_node_socket_addr, client));
-                let mut clients = self.clients.write().unwrap();
-                clients.insert(comit_node_socket_addr, Arc::clone(&client));
-                debug!(
-                    "Client for {} created by making a new connection",
-                    comit_node_socket_addr
+
+                let (sender, receiver) = oneshot::channel();
+                *connection_state = ConnectionState::Connecting(vec![sender]);
+
+                let new_connection_state = Arc::clone(&existing_connection);
+
+                let establish_connection = TcpStream::connect(&socket_addr).then(
+                    move |socket| -> Box<dyn Future<Item = (), Error = ()> + Send> {
+                        info!("Connection to {} established", socket_addr);
```

Comment:
:x: You actually don't know if it is successful here. The log should be moved to the `Ok` match arm below.